### PR TITLE
Kafka on HDInsight

### PR DESCRIPTION
**Pull request by Blackmist**
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use resourceGroup().location for resource locations
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/bp-checklist.md

- [ X ] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

New template to demonstrate how to create a Kafka on HDInsight cluster inside a virtual network.

### CODE REVIEW

File: 101-hdinsight-kafka-vnet/azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+  "parameters": {
+    "clusterName": {
+      "type": "string",
+      "metadata": {
+        "description": "The name of the Kafka cluster to create. This must be a unique name."
+      }
+    },
+    "clusterLoginUserName": {
+      "type": "string",
+      "defaultValue": "admin",
```
Comment (bmoore-msft) [inline 101-hdinsight-kafka-vnet/azuredeploy.json]: remove this - creates surface area

File: 101-hdinsight-kafka-vnet/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    },
+    "clusterLoginPassword": {
+      "type": "securestring",
+      "metadata": {
+        "description": "The password must be at least 10 characters in length and must contain at least one digit, one non-alphanumeric character, and one upper or lower case letter."
+      }
+    },
+    "sshUserName": {
+      "type": "string",
+      "defaultValue": "sshuser",
```
Comment (bmoore-msft) [inline 101-hdinsight-kafka-vnet/azuredeploy.json]: remove this creates surface area

File: 101-hdinsight-kafka-vnet/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+            }
+          }
+        },
+        "storageProfile": {
+          "storageaccounts": [
+            {
+              "name": "[replace(replace(concat(reference(concat('Microsoft.Storage/storageAccounts/', variables('defaultStorageAccount').name), '2016-01-01').primaryEndpoints.blob),'https:',''),'/','')]",
+              "isDefault": true,
+              "container": "[parameters('clusterName')]",
+              "key": "[listKeys(resourceId('Microsoft.Storage/storageAccounts', variables('defaultStorageAccount').name), '2016-01-01').keys[0].value]"
```
Comment (bmoore-msft) [inline 101-hdinsight-kafka-vnet/azuredeploy.json]: update apiVersion

File: 101-hdinsight-kafka-vnet/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+              },
+              "osProfile": {
+                "linuxOperatingSystemProfile": {
+                  "username": "[parameters('sshUserName')]",
+                  "password": "[parameters('sshPassword')]"
+                }
+              },
+              "virtualNetworkProfile": {
+                "id": "[resourceId('Microsoft.Network/virtualNetworks', variables('networkConfig').name)]",
+                "subnet": "[concat(resourceId('Microsoft.Network/virtualNetworks', variables('networkConfig').name), '/subnets/', variables('subnetConfig').name)]"
```
Comment (bmoore-msft) [inline 101-hdinsight-kafka-vnet/azuredeploy.json]: simplify, just use resourceId()

resourceId('Microsoft.Network/virtualNetworks/subnets', variables('networkConfig').name), variables('subnetConfig').name)

File: 101-hdinsight-kafka-vnet/azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+              ],
+              "osProfile": {
+                "linuxOperatingSystemProfile": {
+                  "username": "[parameters('sshUserName')]",
+                  "password": "[parameters('sshPassword')]"
+                }
+              },
+              "virtualNetworkProfile": {
+                "id": "[resourceId('Microsoft.Network/virtualNetworks', variables('networkConfig').name)]",
+                "subnet": "[concat(resourceId('Microsoft.Network/virtualNetworks', variables('networkConfig').name), '/subnets/', variables('subnetConfig').name)]"
```
Comment (bmoore-msft) [inline 101-hdinsight-kafka-vnet/azuredeploy.json]: just use resourceId()

File: 101-hdinsight-kafka-vnet/azuredeploy.json [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+              },
+              "osProfile": {
+                "linuxOperatingSystemProfile": {
+                  "username": "[parameters('sshUserName')]",
+                  "password": "[parameters('sshPassword')]"
+                }
+              },
+              "virtualNetworkProfile": {
+                "id": "[resourceId('Microsoft.Network/virtualNetworks', variables('networkConfig').name)]",
+                "subnet": "[concat(resourceId('Microsoft.Network/virtualNetworks', variables('networkConfig').name), '/subnets/', variables('subnetConfig').name)]"
```
Comment (bmoore-msft) [inline 101-hdinsight-kafka-vnet/azuredeploy.json]: simplify, just use resourceId()

File: 101-hdinsight-kafka-vnet/azuredeploy.json (hunk 1 above)
Comment (Blackmist) [reply 101-hdinsight-kafka-vnet/azuredeploy.json]: fixed

File: 101-hdinsight-kafka-vnet/azuredeploy.json (hunk 2 above)
Comment (Blackmist) [reply 101-hdinsight-kafka-vnet/azuredeploy.json]: fixed

File: 101-hdinsight-kafka-vnet/azuredeploy.json (hunk 3 above)
Comment (Blackmist) [reply 101-hdinsight-kafka-vnet/azuredeploy.json]: fixed

File: 101-hdinsight-kafka-vnet/azuredeploy.json (hunk 4 above)
Comment (Blackmist) [reply 101-hdinsight-kafka-vnet/azuredeploy.json]: Removed the vnet bits

File: 101-hdinsight-kafka-vnet/azuredeploy.json (hunk 5 above)
Comment (Blackmist) [reply 101-hdinsight-kafka-vnet/azuredeploy.json]: Removed the vnet bits

File: 101-hdinsight-kafka-vnet/azuredeploy.json (hunk 6 above)
Comment (Blackmist) [reply 101-hdinsight-kafka-vnet/azuredeploy.json]: Removed the vnet bits

**Pull request merged**